go/tools/gopackagesdriver: add automatic target detection

steeve opened the pull request:
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

This PR introduces automatic target detection so that no user input
is required after the `GOPACKAGESDRIVER` setup. This effectively
deprecates the following environment variables:
- `GOPACKAGESDRIVER_BAZEL_TARGETS`
- `GOPACKAGESDRIVER_BAZEL_QUERY`
- `GOPACKAGESDRIVER_BAZEL_TAG_FILTERS`

It works as follows:
- for `importpath` queries, it will `bazel query` a matching `go_library`
    matching it
- for `file=` queries, it will try to find the matching `go_library` in the
    same package
- since it supports multiple queries at the same time, it will `union` those
    queries and query that

Once the `go_library` targets are found, it will try to build them directly,
which should dramatically speed up compilation times, at the loss of transition
support (which wasn't used as much as I thought it would be).

I may reintroduce it in the future via a user-defined flag (to only build the
part of the graph that needs building).

In any case, toolchain or platforms can be switched with the following
environment variables it configuration transitions are needed:
- `GOPACKAGESDRIVER_BAZEL_FLAGS` which will be passed to `bazel` invocations
- `GOPACKAGESDRIVER_BAZEL_QUERY_FLAGS` which will be passed to `bazel query`
    invocations
- `GOPACKAGESDRIVER_BAZEL_QUERY_SCOPE` which specifies the scope for `importpath` queries (since `gopls` only issues `file=` queries, so **use if you know what you're doing!**)
- `GOPACKAGESDRIVER_BAZEL_BUILD_FLAGS` which will be passed to `bazel build`
    invocations

Finally, the driver will not fail in case of a build failure, and even so
uses `--keep_going` and will return whatever packages that did build.

**Which issues(s) does this PR fix?**

It should fix most of the issues from #2858.

**Other notes for review**
I have added a `.vscode` folder for folks who work on `rules_go` with it so that `gopls` now works there too!

linzhp commented on go/tools/gopackagesdriver/bazel_json_builder.go:
```diff
@@ -17,72 +17,104 @@ package main
+}
+
+func (b *BazelJSONBuilder) packageQuery(importPath string) string {
+	return fmt.Sprintf(`some(kind("go_library", attr(importpath, "%s", //...)))`, importPath)
```
Queries like this would take a very long time to finish in a repository with 100K targets and 2000 third-party libraries, like Uber's Go monorepo

steeve replied:
several things we can do:
1. have the scope configurable via environment
2. restrict the scope to the file's package, which is most of the time where the file lives
3. for `importpath` queries, this is trickier as the package could be anywhere in `//...`, however, package queries don't happen when using `gopls`, as it only gives out `file=` ones

steeve replied:
actually `file=` queries use `same_pkg_direct_rdeps` so it shouldn't take long when editing

steeve replied:
ah wait! it's the universe scope of `//...` that loads package anyway. let me try something else.

linzhp reviewed (commented):
Can we continue to support `GOPACKAGESDRIVER_BAZEL_QUERY` etc? This is important for large repos like Uber's Go monorepo. Although it's huge, most users only work on small subset of targets and only need to index their projects and dependencies. It's quite a waste of time for them to wait for hours for package driver to load `//...`

steeve commented:
@linzhp I have updated the PR, it shouldn't load the whole `//...` for file queries now

steeve commented:
@linzhp I have added `GOPACKAGESDRIVER_BAZEL_QUERY_SCOPE` which should control the scope when issuing `importpath` queries. That said, you should _not_ need it for `gopls` as it issues `file=` queries, and I have removed the whole universe loading for those.

linzhp commented on go/tools/gopackagesdriver/bazel_json_builder.go:
```diff
@@ -17,72 +17,102 @@ package main
+}
+
+func (b *BazelJSONBuilder) packageQuery(importPath string) string {
+	return fmt.Sprintf(`some(kind("go_library", attr(importpath, "%s", %s)))`, importPath, bazelQueryScope)
```
Should be:
```suggestion
	return fmt.Sprintf(`some(kind("go_library", attr(importpath, "%s", deps(%s))))`, importPath, bazelQueryScope)
```
so we can get package information from the dependencies

linzhp commented on go/tools/gopackagesdriver/bazel_json_builder.go:
```diff
@@ -17,72 +17,102 @@ package main
+	ret := make([]string, 0, len(requests))
+	for _, request := range requests {
+		if request == "." || request == "./..." {
+			return wildcardQuery
```
gopls sends a `./...` request to package driver when VS Code opens. This wildcardQuery will also load the universe. Maybe we should just load only `bazelQueryScope`?

```suggestion
			return fmt.Sprintf("kind(go_library, %s)", bazelQueryScope)
```

steeve replied:
that's a very good find.
although i'm thinking it may be best to disable `./...` altogether ?

steeve replied:
my reasoning being, is there a valid usage to querying `./...` with `gopls`? 
I think your change may be the most sensible one, perhaps also allowing folks to specify an empty scope to disable it altogether

linzhp replied:
I am fine with either approach. I am not familiar with gopls enough to say whether `./...` is necessary. Disabling `./...` may make gopls faster when VS code first opens

steeve replied:
I think I found a good workaround: by default `./...` will return only stdlib, unless a scope is specified, in which case it will use that. Same for `importpath` queries.

linzhp replied:
That's actually the same as ignoring `./...` because gopls sends `./... builtin` request when VS Code starts.

steeve commented:
@linzhp I have fixed lots of bugs, can you try this version?

steeve commented:
Here's a summary of the changes:
- when called without `GOPACKAGESDRIVER_BAZEL_QUERY_SCOPE`, return stdlib packages by default (so that vscode doesn't complain)
- add a compiled sources output group to ensure all files are generated and present in the sandbox
- enable querying by `importpath` prefix (`github.com/bazelbuild/rules_go/...`)
- fix path for external generated files
- fix path for external source files

linzhp commented on go/tools/gopackagesdriver/bazel.go:
```diff
@@ -88,7 +103,13 @@ func (b *Bazel) Build(ctx context.Context, args ...string) ([]string, error) {
+		// See https://docs.bazel.build/versions/main/guide.html#what-exit-code-will-i-get on
+		// exit codes.
+		var exerr *exec.ExitError
+		if !errors.As(err, &exerr) || (errors.As(err, &exerr) && exerr.ExitCode() != 1) {
```
equivalent to
```suggestion
		if !errors.As(err, &exerr) || exerr.ExitCode() != 1 {
```

linzhp commented on go/tools/gopackagesdriver/bazel_json_builder.go:
```diff
@@ -17,80 +17,115 @@ package main
+
+func (b *BazelJSONBuilder) packageQuery(importPath string) string {
+	if strings.HasSuffix(importPath, "/...") {
+		importPath = strings.TrimSuffix(importPath, "/...") + "/.*"
```
Should we transform `example.com/pkg/...` to `^example.com/pkg(/.+)?$`, so it includes `example.com/pkg` and `example.com/pkg/foo`, but not `myexample.com/pkg/foo` or `example.com/pkg2`?

steeve replied:
good call

steeve replied:
good call, i did not yet

linzhp commented on go/tools/gopackagesdriver/bazel_json_builder.go:
```diff
@@ -17,80 +17,115 @@ package main
+		} else if strings.HasPrefix(request, "file=") {
+			f := strings.TrimPrefix(request, "file=")
+			ret = append(ret, b.fileQuery(f))
+		} else if bazelQueryScope != "" {
```
is `bazelQueryScope` mandatory to accept package queries?

steeve replied:
i'd like it to, because else it's very easy to shoot yourself in the foot

linzhp commented on go/tools/gopackagesdriver/bazel_json_builder.go:
```diff
@@ -17,80 +17,115 @@ package main
+	if strings.HasSuffix(importPath, "/...") {
+		importPath = strings.TrimSuffix(importPath, "/...") + "/.*"
+	}
+	return fmt.Sprintf(`some(kind("go_library", attr(importpath, "%s", deps(%s))))`, importPath, bazelQueryScope)
```
```suggestion
	return fmt.Sprintf(`kind("go_library", attr(importpath, "%s", deps(%s)))`, importPath, bazelQueryScope)
```
Removing the `some` function would allow us to handle unusual cases nicely:
* When there are more than one `go_library` matching the query, we can return the first one
* When there is no such `go_library`, the `some` function would lead to errors like `ERROR: Evaluation of query failed: argument set is empty`, with exit code 7. It is unclear whether the query has a bug or there is no matching package. If we don't have the `some` function, Bazel query will exit normally with empty result. We can report a better error to gopls for this case.

Same applies to the file query

steeve replied:
good call

linzhp commented on go/tools/gopackagesdriver/bazel_json_builder.go:
```diff
@@ -17,80 +17,115 @@ package main
-		}
-		buildsArgs = append(buildsArgs, queryTargets...)
+func (b *BazelJSONBuilder) Build(ctx context.Context, mode LoadMode) ([]string, error) {
+	labels, err := b.query(ctx, b.queryFromRequests(b.requests...))
```
Need to check whether `len(labels) == 0`

linzhp commented on go/tools/gopackagesdriver/main.go:
```diff
@@ -45,77 +49,64 @@ type driverResponse struct {
+	if err != nil {
 		fmt.Fprintf(os.Stderr, "error: %v", err)
-		os.Exit(1)
+		os.Exit(0)
```
Why exiting with 0?

steeve replied:
because gopls will check the packages driver exit code, and if there is an error, and fall back to `go list`: https://github.com/golang/tools/blob/8c34cc9cafff8f13c3f3ed95a11de309752c1c60/go/packages/external.go#L88-L90

linzhp reviewed (commented):
It works now. Commented on other minor issues.

steeve commented:
@linzhp can you confirm the CLA please?

thinkiny commented:
can't wait this merge to master

linzhp reviewed (commented):
Some tests are failing.

It's strange that googlebot thought I didn't have CLA, given I already had several accepted pull requests to this repo.

@googlebot I consent.

steeve commented:
The test failure is unrelated (race)

steeve commented:
@linzhp I have added https://github.com/bazelbuild/rules_go/pull/2932/commits/3512e5976cf38889c4a94096fe95d4ac3a038cd7 which makes the gopackagesdriver honor the build tags when listing files. Can you try it ?

steeve commented:
> can't wait this merge to master

very soon !

linzhp commented on go/tools/gopackagesdriver/bazel_json_builder.go:
```diff
@@ -17,80 +17,119 @@ package main
+		fp, _ := filepath.Rel(b.bazel.WorkspaceRoot(), filename)
+		filename = fp
+	}
+	return fmt.Sprintf(`kind("go_library", same_pkg_direct_rdeps("%s"))`, filename)
```
This may return empty result when the `filename` is a test file like `foo_test.go`, because no `go_library` would depend on such file

steeve replied:
yeah, test files are not done atm, if you have an idea, i'm all ears

linzhp commented on go/tools/gopackagesdriver/bazel_json_builder.go:
```diff
@@ -17,80 +17,119 @@ package main
+func (b *BazelJSONBuilder) queryFromRequests(requests ...string) string {
+	ret := make([]string, 0, len(requests))
+	for _, request := range requests {
+		if request == "." || request == "./..." {
+			if bazelQueryScope != "" {
+				ret = append(ret, fmt.Sprintf(`kind("go_library", %s)`, bazelQueryScope))
+			} else {
+				ret = append(ret, fmt.Sprintf(RulesGoStdlibLabel))
+			}
+		} else if request == "builtin" || request == "std" {
+			ret = append(ret, fmt.Sprintf(RulesGoStdlibLabel))
+		} else if strings.HasPrefix(request, "file=") {
+			f := strings.TrimPrefix(request, "file=")
+			ret = append(ret, b.fileQuery(f))
+		} else if bazelQueryScope != "" {
+			ret = append(ret, b.packageQuery(request))
+		}
```
```suggestion
		result := ""
		if request == "." || request == "./..." {
			if bazelQueryScope != "" {
				result = fmt.Sprintf(`kind("go_library", %s)`, bazelQueryScope)
			} else {
				result = fmt.Sprintf(RulesGoStdlibLabel)
			}
		} else if request == "builtin" || request == "std" {
			result = fmt.Sprintf(RulesGoStdlibLabel)
		} else if strings.HasPrefix(request, "file=") {
			f := strings.TrimPrefix(request, "file=")
			result = b.fileQuery(f)
		} else if bazelQueryScope != "" {
			result = b.packageQuery(request)
		}
		if result != "" {
			ret = append(ret, result)
		}
```

Need to check whether the query result is empty before appending, otherwise later `bazel build` may complain
```
ERROR: Skipping '': the empty string is not a valid target
WARNING: Target pattern parsing failed.
```

linzhp commented:
> I have added 3512e59 which makes the gopackagesdriver honor the build tags when listing files. Can you try it ?
How should I try it? It seems to be a separate feature. Can you put it in a separate pull request?

linzhp commented:
> The test failure is unrelated (race)

It failed consistently in recent commits, but not in earlier commits or master. So it may be broken by some recent commit

steeve commented:
> How should I try it? It seems to be a separate feature. Can you put it in a separate pull request?

Just verify that it correctly work as it did before. It's done behind the scenes.

steeve commented:
@linzhp careful:
> leaving a comment that **contains only** `@googlebot I consent`. in this pull request.

steeve commented:
Jesus everytime I delete the googlebot mesasge (so that only the last remains), it adds another :(

steeve commented:
I'm _this_ close to making tests work too. The only issue is I can't get the `_test.go` files from within `GoArchive` provider as they are embedded in the `go_test` target, which then compiles them with the `go_default_library`.

linzhp commented:
@googlebot I consent.

linzhp commented:
package driver is now broken:

```
Error loading workspace: /Users/zplin/gocode/src/github.com/bazelbuild/rules_go/tools/gopackagesdriver.sh: exit status 127: /Users/zplin/gocode/src/github.com/bazelbuild/rules_go/tools/gopackagesdriver.sh: line 3: exec: bazel: not found : packages.Load error
```

steeve commented:
I think this is your machine @linzhp `line 3: exec: bazel: not found`, it can't find `bazel`

steeve commented:
So I got tests working in this latest version.

linzhp commented on go/tools/gopackagesdriver/build_context.go:
```diff
@@ -0,0 +1,38 @@
+var buildContext = makeBuildContext()
+
+func makeBuildContext() *build.Context {
+	bctx := &build.Context{
```
I suspect these lines made package driver not able to find Bazel in PATH.

steeve commented:
@linzhp standalone tests are fixed

linzhp commented on go/tools/gopackagesdriver/aspect.bzl:
```diff
@@ -17,90 +17,123 @@ load(
+        # do that for direct dependencies that are not defined as deps, and use the
+        # importpath to find which.
+        if ctx.rule.kind == "go_test":
+            deps_targets = [
```
Maybe it's better to find the `go_library` target in the same directory as the `go_test` with regular (not generated) src files. When Gazelle generate `go_test` without `embed`, it puts the `:go_default_library` in `deps`. So we need that from `deps`.

steeve replied:
I don't think it is, as I try to let each target generate their own files, so that the analysis graph remains stable and the resource usage is kept "low".
If the `go_library` is in `deps`, then things will work as expected without all this code. The hard part is handling `embed`. Consuming the "augmented" version requires consuming `GoArchiveData.direct`, so we can't use the `Targets`. Hence this piece.

thinkiny commented:
awesome
